winch/codegen/src/isa/x64/masm.rs

```diff
@@ -418,6 +418,60 @@ impl Masm for MacroAssembler {
     fn jmp(&mut self, target: MachLabel) {
         self.asm.jmp(target);
     }
+
+    fn popcnt(&mut self, context: &mut CodeGenContext, size: OperandSize) {
+        let src = context.pop_to_reg(self, None, size);
+        if self.flags.has_popcnt() {
+            self.asm.popcnt(src, size);
+            context.stack.push(Val::reg(src));
+        } else {
+            let tmp = context.any_gpr(self);
+            let dst = src;
+            let (masks, shift_amt) = match size {
+                OperandSize::S64 => (
+                    [
+                        0x5555555555555555, // m1
+                        0x3333333333333333, // m2
+                        0x0f0f0f0f0f0f0f0f, // m4
+                        0x0101010101010101, // h01
+                    ],
+                    56u8,
+                ),
+                // 32-bit popcount is the same, except the masks are half as
+                // wide and we shift by 24 at the end rather than 56
+                OperandSize::S32 => (
+                    [0x55555555i64, 0x33333333i64, 0x0f0f0f0fi64, 0x01010101i64],
+                    24u8,
+                ),
+            };
+            self.asm.mov_rr(src, tmp, size);
+
+            // x -= (x >> 1) & m1;
+            self.asm.shift_ir(1u8, dst, ShiftKind::ShrU, size);
+            self.asm.and(RegImm::imm(masks[0]).into(), dst.into(), size);
```

Comment:
I was pretty confused about why you didn't use `and_ir`, until I dug into the implementation enough to understand that x86 only supports 32-bit immediates for `and`, so doing it this way lets the underlying assembler decide whether to use `load_constant` into the scratch register or to emit the immediate operand inline.

I don't know what to suggest but maybe there's a comment that could go here. Or maybe we should just assume that people reading this are more familiar with Winch idioms than I am. :laughing:

Also, it's a little unfortunate that in the 64-bit case, `masks[1]` gets loaded into the scratch register twice. It would be nice to avoid that, but it would clutter the implementation here and maybe that's not worth doing.

Reply:
>  Or maybe we should just assume that people reading this are more familiar with Winch idioms than I am. 😆

The scratch register has caused confusion, I agree. This is something that I'd like to improve, so I'm considering adding guards, which will hopefully make its usage traceable and more explicit.

Reply:
I just haven't gotten to it 😄

Reply:
I don't think it would be terrible to unconditionally call `load_constant` here to put the `0x333...` mask in the scratch register, and use `and_rr` explicitly. It's one extra instruction in the 32-bit case, but it's only a move-immediate, and the generated code is shorter due to only emitting the constant once.